File: src/utils/riskEngine.js
// Risk scoring pipeline with configurable weights and vulnerability adjustments

/**
 * Generate flood risk score for a ward
 * @param {number} rainfall - Rainfall in mm
 * @param {string} timeHorizon - "30m", "1h", or "3h"
 * @param {Object} wardData - Ward-specific data (elevation, drainBlockage, vulnerability)
 * @returns {Object} Risk assessment with score, level, confidence, and action
 */
export function generateRisk(rainfall, timeHorizon, wardData = {}) {
  // Default ward data if not provided
  const {
    elevationFactor = Math.random() * 40,
    drainBlockage = Math.random() * 60,
    vulnerabilityIndex = 1.0,
    populationDensity = 0.5,
    hasHospitals = false,
    hasSchools = false,
    hasMetro = false,
    wardId = null
  } = wardData;

  // Configurable weights per ward (can be tuned based on local conditions)
  const w1 = 0.5; // Rainfall weight
  const w2 = 0.2; // Elevation weight
  const w3 = 0.3; // Drain blockage weight

  // Base risk calculation
  const baseRisk = 
    (rainfall * w1) + 
    (elevationFactor * w2) + 
    (drainBlockage * w3);

  // Time-based multiplier
  const timeMultiplier = 
    timeHorizon === "30m" ? 0.9 :
    timeHorizon === "1h" ? 1.0 : 
    timeHorizon === "3h" ? 1.15 : 1.0;

  // Vulnerability adjustment
  let vulnerabilityMultiplier = vulnerabilityIndex;
  
  // Increase risk if critical infrastructure present
  if (hasHospitals) vulnerabilityMultiplier += 0.15;
  if (hasSchools) vulnerabilityMultiplier += 0.1;
  if (hasMetro) vulnerabilityMultiplier += 0.1;
  if (populationDensity > 0.7) vulnerabilityMultiplier += 0.1;

  // Final risk score
  const finalRisk = Math.min(100, Math.round(baseRisk * timeMultiplier * vulnerabilityMultiplier));

  // Confidence calculation (higher for shorter time horizons and consistent data)
  let confidence = 85;
  if (timeHorizon === "3h") confidence -= 10;
  if (timeHorizon === "30m") confidence += 5;
  
  // Reduce confidence if data is incomplete
  if (!wardId) confidence -= 5;
  
  confidence = Math.max(60, Math.min(95, confidence + (Math.random() * 10 - 5)));

  // Risk level classification
  let level = "Low";
  let action = "Monitoring only";
  let priority = "low";

  if (finalRisk >= 70) {
    level = "Critical";
    priority = "critical";
    action = "Deploy 2 pumps & barricade roads immediately";
  } else if (finalRisk >= 50) {
    level = "High";
    priority = "high";
    action = "Deploy 1 pump, inspect drains, prepare barricades";
  } else if (finalRisk >= 30) {
    level = "Medium";
    priority = "medium";
    action = "Inspect drains, keep pumps on standby";
  }

  return {
    score: finalRisk,
    level,
    confidence: Math.round(confidence),
    action,
    priority,
    rainfall,
    drainBlockage: Math.round(drainBlockage),
    elevationFactor: Math.round(elevationFactor),
    vulnerabilityIndex: vulnerabilityMultiplier.toFixed(2),
    timeHorizon
  };
}

/**
 * Generate risk for multiple wards
 */
export function generateWardRisks(rainfall, timeHorizon, wards) {
  return wards.map(ward => {
    const wardData = {
      wardId: ward.properties?.ward_id || ward.properties?.id,
      elevationFactor: ward.properties?.elevation || Math.random() * 40,
      drainBlockage: ward.properties?.drain_blockage || Math.random() * 60,
      vulnerabilityIndex: ward.properties?.vulnerability || 1.0,
      populationDensity: ward.properties?.population_density || Math.random(),
      hasHospitals: ward.properties?.has_hospitals || Math.random() > 0.7,
      hasSchools: ward.properties?.has_schools || Math.random() > 0.5,
      hasMetro: ward.properties?.has_metro || Math.random() > 0.8,
    };
    
    return {
      ...ward,
      risk: generateRisk(rainfall, timeHorizon, wardData)
    };
  });
}

/**
 * Calculate city-wide preparedness score
 */
export function calculatePreparednessScore(wardRisks) {
  if (!wardRisks || wardRisks.length === 0) return 0;
  
  const totalWards = wardRisks.length;
  const criticalWards = wardRisks.filter(w => w.risk?.level === "Critical").length;
  const highWards = wardRisks.filter(w => w.risk?.level === "High").length;
  const mediumWards = wardRisks.filter(w => w.risk?.level === "Medium").length;
  
  // Preparedness decreases with more high-risk wards
  const riskPenalty = (criticalWards * 20) + (highWards * 10) + (mediumWards * 5);
  const preparedness = Math.max(0, Math.min(100, 100 - (riskPenalty / totalWards)));
  
  return Math.round(preparedness);
}

/**
 * Generate task tickets from high-risk wards
 */
export function generateTaskTickets(wardRisks) {
  const tickets = [];
  
  wardRisks.forEach(ward => {
    const risk = ward.risk;
    if (risk && (risk.level === "Critical" || risk.level === "High")) {
      const wardName = ward.properties?.name || `Ward ${ward.properties?.ward_id || 'Unknown'}`;
      
      tickets.push({
        id: `ticket-${ward.properties?.ward_id || Date.now()}-${Math.random()}`,
        wardName,
        location: wardName,
        severity: risk.level,
        priority: risk.priority,
        action: risk.action,
        timeWindow: risk.timeHorizon === "30m" ? "30 minutes" : 
                   risk.timeHorizon === "1h" ? "1 hour" : "3 hours",
        riskScore: risk.score,
        confidence: risk.confidence,
        timestamp: new Date().toISOString()
      });
    }
  });
  
  // Sort by priority (critical first, then by risk score)
  return tickets.sort((a, b) => {
    if (a.severity === "Critical" && b.severity !== "Critical") return -1;
    if (a.severity !== "Critical" && b.severity === "Critical") return 1;
    return b.riskScore - a.riskScore;
  });
}

File: backend/risk_engine.py
"""
Risk Engine for Flood Prediction
Implements the risk scoring pipeline with configurable weights
"""

import random
from typing import Optional, Dict, List

def calculate_risk_score(
    rainfall: float,
    time_horizon: str,
    ward_data: Optional[Dict] = None
) -> Dict:
    """
    Calculate flood risk score for a ward
    
    Args:
        rainfall: Rainfall in mm
        time_horizon: "30m", "1h", or "3h"
        ward_data: Optional ward-specific data
    
    Returns:
        Dictionary with risk assessment
    """
    # Default ward data if not provided
    if ward_data is None:
        ward_data = {}
    
    elevation_factor = ward_data.get("elevation_factor", random.random() * 40)
    drain_blockage = ward_data.get("drain_blockage", random.random() * 60)
    vulnerability_index = ward_data.get("vulnerability_index", 1.0)
    population_density = ward_data.get("population_density", 0.5)
    has_hospitals = ward_data.get("has_hospitals", False)
    has_schools = ward_data.get("has_schools", False)
    has_metro = ward_data.get("has_metro", False)
    
    # Configurable weights
    w1 = 0.5  # Rainfall weight
    w2 = 0.2  # Elevation weight
    w3 = 0.3  # Drain blockage weight
    
    # Base risk calculation
    base_risk = (
        (rainfall * w1) +
        (elevation_factor * w2) +
        (drain_blockage * w3)
    )
    
    # Time-based multiplier
    time_multiplier = {
        "30m": 0.9,
        "1h": 1.0,
        "3h": 1.15
    }.get(time_horizon, 1.0)
    
    # Vulnerability adjustment
    vulnerability_multiplier = vulnerability_index
    
    # Increase risk if critical infrastructure present
    if has_hospitals:
        vulnerability_multiplier += 0.15
    if has_schools:
        vulnerability_multiplier += 0.1
    if has_metro:
        vulnerability_multiplier += 0.1
    if population_density > 0.7:
        vulnerability_multiplier += 0.1
    
    # Final risk score
    final_risk = min(100, round(base_risk * time_multiplier * vulnerability_multiplier))
    
    # Confidence calculation
    confidence = 85
    if time_horizon == "3h":
        confidence -= 10
    elif time_horizon == "30m":
        confidence += 5
    
    if not ward_data.get("ward_id"):
        confidence -= 5
    
    confidence = max(60, min(95, confidence + (random.random() * 10 - 5)))
    
    # Risk level classification
    level = "Low"
    action = "Monitoring only"
    priority = "low"
    
    if final_risk >= 70:
        level = "Critical"
        priority = "critical"
        action = "Deploy 2 pumps & barricade roads immediately"
    elif final_risk >= 50:
        level = "High"
        priority = "high"
        action = "Deploy 1 pump, inspect drains, prepare barricades"
    elif final_risk >= 30:
        level = "Medium"
        priority = "medium"
        action = "Inspect drains, keep pumps on standby"
    
    return {
        "score": final_risk,
        "level": level,
        "confidence": round(confidence),
        "action": action,
        "priority": priority,
        "rainfall": rainfall,
        "drain_blockage": round(drain_blockage),
        "elevation_factor": round(elevation_factor),
        "vulnerability_index": f"{vulnerability_multiplier:.2f}",
        "time_horizon": time_horizon
    }

def generate_ward_risks(
    rainfall: float,
    time_horizon: str,
    wards: List[Dict]
) -> List[Dict]:
    """
    Generate risk assessments for multiple wards
    """
    ward_risks = []
    
    for ward in wards:
        props = ward.get("properties", {})
        ward_data = {
            "ward_id": props.get("ward_id") or props.get("id"),
            "elevation_factor": props.get("elevation", random.random() * 40),
            "drain_blockage": props.get("drain_blockage", random.random() * 60),
            "vulnerability_index": props.get("vulnerability", 1.0),
            "population_density": props.get("population_density", random.random()),
            "has_hospitals": props.get("has_hospitals", random.random() > 0.7),
            "has_schools": props.get("has_schools", random.random() > 0.5),
            "has_metro": props.get("has_metro", random.random() > 0.8),
        }
        
        risk = calculate_risk_score(rainfall, time_horizon, ward_data)
        
        ward_risks.append({
            **ward,
            "risk": risk
        })
    
    return ward_risks

def calculate_preparedness(ward_risks: List[Dict]) -> Dict:
    """
    Calculate city-wide preparedness score
    """
    if not ward_risks:
        return {
            "score": 0,
            "label": "Unknown",
            "total_wards": 0,
            "critical_wards": 0,
            "high_wards": 0,
            "medium_wards": 0,
            "low_wards": 0
        }
    
    total_wards = len(ward_risks)
    critical_wards = sum(1 for w in ward_risks if w.get("risk", {}).get("level") == "Critical")
    high_wards = sum(1 for w in ward_risks if w.get("risk", {}).get("level") == "High")
    medium_wards = sum(1 for w in ward_risks if w.get("risk", {}).get("level") == "Medium")
    low_wards = sum(1 for w in ward_risks if w.get("risk", {}).get("level") == "Low")
    
    # Preparedness decreases with more high-risk wards
    risk_penalty = (critical_wards * 20) + (high_wards * 10) + (medium_wards * 5)
    preparedness = max(0, min(100, 100 - (risk_penalty / total_wards)))
    
    score = round(preparedness)
    
    if score >= 80:
        label = "Excellent"
    elif score >= 60:
        label = "Good"
    elif score >= 40:
        label = "Fair"
    else:
        label = "Critical"
    
    return {
        "score": score,
        "label": label,
        "total_wards": total_wards,
        "critical_wards": critical_wards,
        "high_wards": high_wards,
        "medium_wards": medium_wards,
        "low_wards": low_wards
    }

